plugins/askrene/explain_failure.c

```diff
@@ -0,0 +1,290 @@
+		}
+	}
+
+	return tal_fmt(ctx, "marked disabled by gossip message.");
+}
+
+static const char *describe_capacity(const tal_t *ctx,
+				     const struct route_query *rq,
+				     const struct short_channel_id_dir *scidd,
+				     struct amount_msat amount)
+{
+	for (int i = tal_count(rq->layers) - 1; i >= 0; i--) {
+		if (layer_created(rq->layers[i], scidd->scid)) {
+			return tal_fmt(ctx, " (created by layer %s) isn't big enough to carry %s.",
+				       layer_name(rq->layers[i]),
+				       fmt_amount_msat(tmpctx, amount));
+		}
+	}
+
+	return tal_fmt(ctx, "isn't big enough to carry %s.",
+		       fmt_amount_msat(tmpctx, amount));
+}
+
+/* We failed to find a flow at all.  Why? */
+const char *explain_failure(const tal_t *ctx,
+			    const struct route_query *rq,
+			    const struct gossmap_node *srcnode,
+			    const struct gossmap_node *dstnode,
+			    struct amount_msat amount)
+{
+	const struct route_hop *hops;
+	const struct dijkstra *dij;
+	char *path;
+	const char *cap_check;
+
+	/* Do we have enough funds? */
+	cap_check = check_capacity(ctx, rq, srcnode, OUT_OF_NODE,
+				   amount, "source");
+	if (cap_check)
+		return cap_check;
+
+	/* Does destination have enough capacity? */
+	cap_check = check_capacity(ctx, rq, dstnode, INTO_NODE,
+				   amount, "destination");
+	if (cap_check)
+		return cap_check;
+
+	/* OK, fall back to telling them why didn't shortest path
+	 * work.  This covers the "but I have a direct channel!"
+	 * case. */
+	dij = dijkstra(tmpctx, rq->gossmap, dstnode, AMOUNT_MSAT(0), 0,
+		       always_true, route_score_one, NULL);
+	hops = route_from_dijkstra(tmpctx, rq->gossmap, dij, srcnode,
+				   AMOUNT_MSAT(0), 0);
+	if (!hops)
+		return tal_fmt(ctx, "There is no connection between source and destination at all");
+
+	/* Description of shortest path */
+	path = tal_strdup(tmpctx, "");
+	for (size_t i = 0; i < tal_count(hops); i++) {
+		tal_append_fmt(&path, "%s%s",
+			       i > 0 ? "->" : "",
+			       fmt_short_channel_id(tmpctx, hops[i].scid));
+	}
+
+	/* Now walk through this: is it disabled?  Insuff capacity? */
+	for (size_t i = 0; i < tal_count(hops); i++) {
+		const char *explanation;
+		struct short_channel_id_dir scidd;
+		struct gossmap_chan *c;
+		struct amount_msat cap_msat;
+
+		scidd.scid = hops[i].scid;
+		scidd.dir = hops[i].direction;
+		c = gossmap_find_chan(rq->gossmap, &scidd.scid);
+		cap_msat = gossmap_chan_get_capacity(rq->gossmap, c);
+		if (!gossmap_chan_set(c, scidd.dir))
+			explanation = "has no gossip";
+		else if (!c->half[scidd.dir].enabled)
+			explanation = describe_disabled(tmpctx, rq, &scidd);
+		else if (amount_msat_greater(amount, cap_msat))
+			explanation = describe_capacity(tmpctx, rq, &scidd, amount);
+		else {
+			struct amount_msat min, max;
+			get_constraints(rq, c, scidd.dir, &min, &max);
+			if (amount_msat_less(max, amount)) {
+				explanation = why_max_constrained(tmpctx, rq,
+								  &scidd, amount);
+			} else
+				continue;
+		}
```

Comment:
We can get constraints above and simplify this else statement
```suggestion
		else if (amount_msat_less(max, amount)) {
				explanation = why_max_constrained(tmpctx, rq,
								  &scidd, amount);
		}
```